### ISSUE TITLE
Performance is slow for tables with a lot of columns (and cells)

### ISSUE THREAD

**Issue by afshin** [twosigma/beakerx_tabledisplay#68]
`beakerx_tabledisplay` performance takes a big hit and can freeze the browser when there are a lot of cells.

To reproduce this issue:

## Set up
```
conda create -y -n slowtable
conda activate slowtable
conda install -y jupyterlab
pip install beakerx_tabledisplay
beakerx_tabledisplay install
jupyter lab
```

## Execute

This is the code I ran to manifest the issue. Here is a zipped notebook containing the code below: [slow-tabledisplay.ipynb.zip](https://github.com/twosigma/beakerx_tabledisplay/files/8072517/slow-tabledisplay.ipynb.zip)

### Cell 1
```python
import beakerx_tabledisplay
import pandas as pd
table = pd.DataFrame({k: range(10) for k in range(1000)})
```
### Cell 2
```python
display(table)
```

## Note on JupyterLab 2 vs. 3
Interestingly, while this is slow in **both** JupyterLab 2.x and also JupyterLab 3.x, **it seems slightly *slower* in JupyterLab 3**, which causes this warning to arise in Firefox:

<img width="975" alt="firefox-warning" src="https://user-images.githubusercontent.com/159529/154099450-3887b448-6afc-4cf6-86d6-63f9a7151438.png">

**Comment by davidbrochart** [comment]
I wanted to see how the new kernel protocol over websocket improved the situation, but it is being implemented in JupyterLab 4.0, which is not compatible with this widget.
Is there a development version that would support JupyterLab 4.0?

**Comment by afshin** [comment]
Hi @davidbrochart! This extension doesn't have a 4.x version yet. It's something we should work on soon.

But I don't think the websocket changes will improve this because you can actually save the notebook with all of its outputs in the document and open it without a kernel or any websocket traffic and it'll still be slow. This particular slowness is definitely a front-end phenomenon.

**Comment by davidbrochart** [comment]
Thanks @afshin, I managed to install from source. I'll run some benchmark anyway.

**Comment by davidbrochart** [comment]
I accumulated all traffic over websocket in Jupyter Server, and I can see that the new protocol is 5.5 times faster on this example:
```python
table = pd.DataFrame({k: range(1000) for k in range(1000)})
```
Total websocket traffic:
- old protocol: 0.088 s
- new protocol: 0.016 s

**Comment by afshin** [comment]
That's excellent and I'm glad to see this optimization! But even if the transfer was instantaneous, the front-end chokes trying to render this table.

**Comment by davidbrochart** [comment]
With a bigger table, the new protocol is now 18x faster:
```python
table = pd.DataFrame({k: range(1000) for k in range(10_000)})
```
Total websocket traffic:
- old protocol: 0.8726 s
- new protocol: 0.04865 s

**Comment by davidbrochart** [comment]
In terms of data rate (from ZMQ to WebSocket), I got the following results on a consumer laptop (i7 @1.80 GHz, 16 GB of RAM):
- old protocol: 53 MB/s
- new protocol: 922 MB/s

**Comment by afshin** [comment]
Thanks for investigating this further, @davidbrochart!

Just to clarify, are your results that this slowness is a data transmission / parsing issue? My guess was that this is a client-side performance issue; is that an incorrect assumption?

---
Also, these figures might also be useful as a case study in your [protocol alignment PR in JupyterLab](https://github.com/jupyterlab/jupyterlab/pull/11841).

**Comment by davidbrochart** [comment]
No, this doesn't explain the slowness of the beakerx display, which as you said is a front-end issue. Actually, these benchmarks were done with a table of 10_000_000 elements, which never displays in the notebook, but I can see that the transfer over the websocket is done on the server side.

**Comment by fcollonval** [comment]
Bringing here a comparison of the same table using [ipydatagrid](https://github.com/bloomberg/ipydatagrid) that uses the Lumino DataGrid behind the scene too.

![ipydatagrid_vs_beakerx](https://user-images.githubusercontent.com/8435071/155353966-aea2a680-b334-40e5-9542-06353682ac3d.gif)

So this confirms the lumino datagrid can achieve good performance. Next step profiling the code to figure out the bottleneck.
